fdbkubernetesmonitor/api/version.go

```diff
@@ -0,0 +1,197 @@
+/*
+ * version.go
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2021-2024 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package api
+
+import (
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+)
+
+// Version represents a version of FoundationDB.
+//
+// This provides convenience methods for checking features available in
+// different versions.
+type Version struct {
+	// Major is the major version
+	Major int
+
+	// Minor is the minor version
+	Minor int
+
+	// Patch is the patch version
+	Patch int
+
+	// ReleaseCandidate is the number from the `-rc\d+` suffix version
+	// of the version if it exists
+	ReleaseCandidate int
+}
+
+// versionRegex describes the format of a FoundationDB version.
+var versionRegex = regexp.MustCompile(`(\d+)\.(\d+)\.(\d+)(-rc(\d+))?`)
+
+// MarshalJSON custom implementation of MarshalJSON
+func (version *Version) MarshalJSON() ([]byte, error) {
+	return []byte(fmt.Sprintf("\"%s\"", version.String())), nil
+}
+
+// UnmarshalJSON custom implementation of UnmarshalJSON
+func (version *Version) UnmarshalJSON(data []byte) error {
+	trimmed := strings.Trim(string(data), "\"")
+	currentVersion, err := ParseFdbVersion(trimmed)
+	if err != nil {
+		return err
+	}
+
+	version.Major = currentVersion.Major
+	version.Minor = currentVersion.Minor
+	version.Patch = currentVersion.Patch
+	version.ReleaseCandidate = currentVersion.ReleaseCandidate
+	return nil
+}
+
+// ParseFdbVersion parses a version from its string representation.
+func ParseFdbVersion(version string) (Version, error) {
+	matches := versionRegex.FindStringSubmatch(version)
+	if matches == nil {
```

Comment:
is this guaranteed to have len 6+ if there are no errors? (I think yes based on reading + tests below, but my regex is rather rusty and slice index errors are bad)

Reply:
ah, I see, this is previously used code, even less likely to be a problem...